components/core/src/clp_s/JsonFileIterator.cpp

```diff
@@ -81,58 +83,67 @@ bool JsonFileIterator::read_new_json() {
-            // naturally the truncated bytes for the stream.
-            m_next_document_position = m_doc_it.current_index() + m_doc_it.source().size();
-            return true;
-        } else if (m_doc_it.error() == simdjson::error_code::UTF8_ERROR) {
-            m_truncated_bytes = m_buf_occupied - m_next_document_position;
-        } else {
-            m_error_code = m_doc_it.error();
-            return false;
-        }
-    } else if (m_eof) {
-        // if we iterate to the end of the stream without error after hitting eof it should be safe
-        // to ask the stream iterator for number of truncated bytes
-        m_truncated_bytes = m_stream.truncated_bytes();
-        return false;
-    } else {
-        m_truncated_bytes = m_stream.truncated_bytes();
-    }
-
-    if (m_truncated_bytes > 0) {
-        ++m_doc_it;
-        // UTF8 error wasn't at the end of the stream -- terminate
+    do {
+        bool maybe_utf8_edge_case{false};
         if (m_doc_it != m_stream.end()) {
-            m_error_code = simdjson::error_code::UTF8_ERROR;
-            return false;
+            if (simdjson::error_code::SUCCESS == m_doc_it.error()) {
+                it = m_doc_it;
+
+                // There is a bug in simdjson where truncated bytes isn't set correctly when a UTF8
+                // character is cut off at the end of the buffer. To work around this limitation we
+                // always keep track of the start of the next document ourselves. The bytes after
+                // the last valid document and before the end of the buffer are naturally the
+                // truncated bytes for the current buffer.
+                m_next_document_position = m_doc_it.current_index() + m_doc_it.source().size();
+                return true;
+            } else if (m_doc_it.error() == simdjson::error_code::UTF8_ERROR) {
+                maybe_utf8_edge_case = true;
+            } else {
+                m_error_code = m_doc_it.error();
+                return false;
+            }
+        } else {
+            m_truncated_bytes = skip_whitespace_and_get_truncated_bytes();
+            // If we have reached the end of the file and the end of the document stream there are
+            // no more documents to iterate over
+            if (m_eof) {
+                return false;
+            }
         }
-    }
 
-    if (false == read_new_json()) {
-        return false;
-    }
-
-    if (m_doc_it != m_stream.end()) {
-        if (simdjson::error_code::SUCCESS == m_doc_it.error()) {
-            it = m_doc_it;
-            return true;
+        if (maybe_utf8_edge_case) {
+            // Check if this is the UTF-8 edge case by advancing the iterator and checking if this
+            // is really the end of the stream.
+            ++m_doc_it;
+            if (m_doc_it != m_stream.end()) {
+                // Regular UTF-8 error in the middle of the document stream.
+                m_error_code = simdjson::error_code::UTF8_ERROR;
```

Comment:
Not sure if it's a case of UTF-8 error in the middle of the document stream
```"{\"a\"\xff: 1, \"b\": 2}\n{\"a\": 1, \"b\": 2}\n"```. simdjson reports UTF8_ERROR and on error cases, and `++` just sets `finished` to true which results in `m_doc_it == m_stream.end()`. Can you come up with a case of such regular UTF-8 error in the middle of the document stream?

Reply:
Yeah, I'm worried about the case where a developer creates a string field with invalid UTF-8 character (e.g. they copy the hex for some special character off the web incorrectly). If incrementing the iterator always leads to the end iterator on error we'll have to approach this case differently. Maybe we could add a "maximum document size" configuration option with a default of e.g. 200MB then report UTF8 error if either eof is hit or the buffer size is greater than the maximum document size? (And also introduce another new error for document size being too large?)

Reply:
Yeah, I'm a little bit worried about the memory usage and efficiency. From what I noticed, this if branch cannot catch such errors, and it will keep doubling the buffer size until reaching the end of the file. If we can not distinguish different UTF-8 errors from simdjson, the "maximum document size" configuration option could work.